TripKit/AddOns/InterApp/TKInterAppCommunicator.m

```diff
@@ -81,94 +80,65 @@ + (void)openSegmentInMapsApp:(TKSegment *)segment
-                 launchOptions:options];
-}
-
-+ (BOOL)deviceHasGoogleMaps
-{
-  NSURL *testURL = [NSURL URLWithString:@"comgooglemaps-x-callback://"];
-  return [[UIApplication sharedApplication] canOpenURL:testURL];
+  [TKInterAppCommunicator openAppMapsInMode:mode routeFrom:origin to:[segment end]];
 }
 
 + (void)openSegmentInGoogleMapsApp:(TKSegment *)segment
             currentLocationHandler:(nullable BOOL (^)(TKSegment * __nonnull))currentLocationHandler
 {
-  // https://developers.google.com/maps/documentation/ios/urlscheme
-  
-  NSMutableString *directionsRequest = [NSMutableString stringWithString:@"comgooglemaps-x-callback://?"];
-  
-  // source
+  // origin
+  id<MKAnnotation> origin;
   if (currentLocationHandler == nil || currentLocationHandler(segment)) {
     // nothing to add
   } else {
-    CLLocationCoordinate2D origin = [[segment start] coordinate];
-    [directionsRequest appendFormat:@"saddr=%.5f,%.5f&", origin.latitude, origin.longitude];
+    origin = [segment start];
   }
   
-  // destination
-  CLLocationCoordinate2D destination = [[segment end] coordinate];
-  [directionsRequest appendFormat:@"daddr=%.5f,%.5f&", destination.latitude, destination.longitude];
-  
+  // mode
+  TKInterAppCommunicatorMapDirectionMode mode;
   if ([segment isWalking]) {
-    [directionsRequest appendString:@"directionsmode=walking"];
+    mode = TKInterAppCommunicatorMapDirectionModeWalking;
   } else if ([segment isCycling]) {
-    [directionsRequest appendString:@"directionsmode=bicycling"];
+    mode = TKInterAppCommunicatorMapDirectionModeCycling;
   } else if ([segment isDriving]) {
-    [directionsRequest appendString:@"directionsmode=driving"];
+    mode = TKInterAppCommunicatorMapDirectionModeDriving;
+  } else {
+    mode = TKInterAppCommunicatorMapDirectionModeDefault;
   }
   
-  // call-back
-  NSString *callback = [[SGKConfig sharedInstance] googleMapsCallback];
-  if (callback) {
-    [directionsRequest appendFormat:@"x-success=%@", callback];
-  }
-  NSURL *directionsURL = [NSURL URLWithString:directionsRequest];
-  [[UIApplication sharedApplication] openURL:directionsURL];
-}
-
-+ (BOOL)deviceHasWaze
-{
-  NSURL *testURL = [NSURL URLWithString:@"waze://"];
-  return [[UIApplication sharedApplication] canOpenURL:testURL];
+  [TKInterAppCommunicator openGoogleMapsInMode:mode routeFrom:origin to:[segment end]];
 }
 
 + (void)openSegmentInWazeApp:(TKSegment *)segment
 {
-  // https://www.waze.com/about/dev
-  
   // Waze will always start at the current location
-  CLLocationCoordinate2D destination = [[segment end] coordinate];
-  NSMutableString *directionsRequest = [NSMutableString stringWithFormat:@"waze://?ll=%f,%f&navigate=yes", destination.latitude, destination.longitude];
-  
-  NSURL *directionsURL = [NSURL URLWithString:directionsRequest];
-  [[UIApplication sharedApplication] openURL:directionsURL];
+  if ([segment end] != nil) {
+    [TKInterAppCommunicator openWazeRouteTo:[segment end]];
+  }
```

Comment:
Should there be an assert? It's weird to call `openSegmentInWazeApp` and then nothing happens.